src/main/scala/is/hail/io/vcf/LoadVCF.scala

```diff
@@ -202,6 +375,51 @@ object LoadVCF {
       .toArray
   }
 
+  // parses the Variant (key), leaves the rest to f
+  def parseLines[C](makeContext: () => C)(f: (C, String, RegionValueBuilder) => Unit)(
+    lines: RDD[WithContext[String]], t: Type): RDD[RegionValue] = {
+    lines.mapPartitions { it =>
+      new Iterator[RegionValue] {
```

Comment:
This could be a real iterator if you use a stale pattern. The conditional jump in `next` is present in the current code and using the stale pattern. We effectively use one more bit to remember if we've called `hasNext` but not yet called `next`. There are really three states, but we use two-bits to represent them:

 - `present = true`, `stale = false`, we already iterated enough to find the value, and we're ready to return it
 - `present = false`, `stale = false`, we already iterated enough to know there's nothing left in this iterator
 - `stale = true`, we need to iterate to know if there is more in this iterator and to know if there is a value

If we're throwing away the idea that `hasNext` is idempotent then this doesn't matter. We do add an extra conditional jump in `hasNext`, but if the consumer is using a for-each-like pattern, then the condition is always true in `hasNext` and false in `next`, so maybe branch prediction makes this version near-equivalent?

```scala
new Iterator[RegionValue] {
  val rvb = new RegionValueBuilder(region)
  val rv = RegionValue(region)

  val context: C = makeContext()

  private var present: Boolean = false
  private var stale: Boolean = true

  private def advance() {
    while (!present && it.hasNext) {
      present = it.next().map { l =>
        val vcfLine = new VCFLine(l)
        region.clear()
        rvb.start(t)
        rvb.startStruct()
        val b = vcfLine.parseAddVariant(rvb)
        if (b) {
          f(context, l, rvb)

          rvb.endStruct()
          rv.setOffset(rvb.end())
        } else
          rvb.clear()
        b
      }.value
    }
    stale = false
  }

  def hasNext: Boolean = {
    if (stale)
      advance()
    present
  }

  def next(): RegionValue = {
    if (stale)
      advance()
    stale = true
    rv
  }
}
```